src/python/pants/engine/internals/parametrize.py

```diff
@@ -174,34 +195,32 @@ def expand(
             yield (address, fields)
             return
 
-        non_parametrized = tuple(
-            (field_name, field_value)
-            for field_name, field_value in fields.items()
-            if not isinstance(field_value, Parametrize)
-        )
-
         for parametrized_args in itertools.product(*parametrized):
+            # There will be at most one group per cross product.
+            group_kwargs: Mapping[str, Any] = next(
+                (
+                    field_value.kwargs
+                    for _, _, field_value in parametrized_args
+                    if isinstance(field_value, Parametrize) and field_value.is_group
+                ),
+                {},
+            )
             expanded_address = address.parametrize(
-                {field_name: alias for field_name, alias, _ in parametrized_args}
+                {
+                    field_name: alias
+                    for field_name, alias, _ in parametrized_args
+                    if field_name not in group_kwargs
+                }
             )
-            parametrized_args_fields = tuple(
+            expanded_address = expanded_address.unparametrize(list(group_kwargs.keys()))
```

Comment:
I may be wrong here, but I feel this could be a hot code path, and feel that we risk creating multiple instances of the Address in the course of expanding parametrizations here.. perhaps we could keep a local `parameters` variable during processing, then have a single `address.parametrize(parameters, replace=True)` at the end, and introduce a new argument `replace` that doesn't merge in any pre-existing parameters from the address, rather than calling `unparametrize` unconditionally for all addresses (I hope we are called with all field values here.. but I may be mistaken, in which case that won't work.. will have to take a closer look, just putting my concerns out there for now..)

Reply:
One last question: seems like `address.parametrize` is only used in two spot and both are replacing the existing values. Would it make sense to just support replacing parameters instead of providing the `replace: bool` arg?